src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java

```diff
@@ -593,4 +610,252 @@ private static boolean isTypeField(DetailAST variableDef) {
         return parentType == TokenTypes.CLASS_DEF;
     }
 
+    /**
+     * Parses the files lines and assigns them flags depending on whether the line is code,
+     * comment, and/or empty.
+     *
+     * @return list of flags for each line.
+     */
+    private List<EnumSet<LineFlag>> categorizeLines() {
+        final EmptyLineSeparatorParser parser = new EmptyLineSeparatorParser();
+        parser.processLines(getLines());
+        return parser.getFlags();
+    }
+
+    /**
+     * Flags that apply to source code file lines.
+     */
+    private enum LineFlag {
+
+        /**
+         * Line contains code.
+         */
+        CODE,
+
+        /**
+         * Line contains a comment.
+         */
+        COMMENT,
+
+        /**
+         * Line leaves a comment open.
+         */
+        COMMENT_OPEN,
+
+        /**
+         * Line is empty. It doesn't contain any comments or code, but may contain whitespace.
+         */
+        EMPTY
+    }
+
+    /**
+     * Parses a Java source file to assign each line a set of flags based on whether they have code,
```

Comment:
Why are we parsing the Java source? It seems like you are attempting a crude attempt at understanding the source code when we have the ANTLR to do this for us.
Why can't we use the AST to understand the source?

Reply:
I investigated that route, but it seemed to me that the AST didn't have understanding of whitespace, JavaDoc, comments, and empty lines. If it is possible to change the way the AST is built to do this, the I agree, that would be better. Would you be able to point me in the right direction?

Also, prior to my changes the empty lines were detected by checking the lines with `CommonUtils.isBlank`, not by analysing the AST. I assumed this was because it couldn't be done with the AST.

Reply:
> empty lines were detected by checking the lines with CommonUtils.isBlank, not by analysing the AST

Any empty line is an empty line. The util method is fine for this case. Emptyness could also be determined by looking for holes in the AST where nothing is on a specific line.

> AST didn't have understanding of whitespace, JavaDoc, comments, and empty lines

This check is comment aware, so it knows about comments and javadocs. Comments aren't part of the `getAcceptableTokens` list, so `visitToken` won't stop on them by default but they are still in the tree.
Empty lines can be determined using `CommonUtils.isBlank`. The AST contains the line number.
Whitespace would have to be determined by looking at the line's column position and checking what comes before/after it.

> Would you be able to point me in the right direction?

I can but you would need to ask a more specific question.
Be aware, that comments are placed in weird locations sometimes, so it might be best that you add comments to the acceptable list to stop on them.

Example from issue:
````
    // Method 2 must fail
    public static final void method2() { }

    |--METHOD_DEF -> METHOD_DEF [8:4]
    |  |--MODIFIERS -> MODIFIERS [8:4]
    |  |  |--SINGLE_LINE_COMMENT -> // [7:4]
    |  |  |  `--COMMENT_CONTENT ->  Method 2 must fail\n [7:6]
    |  |  |--LITERAL_PUBLIC -> public [8:4]
    |  |  |--LITERAL_STATIC -> static [8:11]
````

You can use http://checkstyle.sourceforge.net/cmdline.html#Command_line_usage option `-T` to print the AST with the comments to under the tree more.

Reply:
@rnveach: Ok, I can look into this. I wish I had known about that before I had sunk hours into making all the CIs pass.